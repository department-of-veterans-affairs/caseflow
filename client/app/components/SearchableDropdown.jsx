import React, { Component, PropTypes } from 'react';
import Select from 'react-select';

class SearchableDropdown extends Component {

  constructor(props) {
    super(props);
    this.state = {
      value: null
    };
  }

  onChange = (value) => {
    let newValue = value;

    /*
     * this is a temp fix for react-select value backspace
     * issue.
     * Setting value to null when an option is deselected
     * using the backspace.
     * https://github.com/JedWatson/react-select/pull/773
     */

    if (Array.isArray(value) && value.length <= 0) {
      newValue = null;
    }
    this.setState({ value: newValue });
    if (this.props.onChange) {
      this.props.onChange(newValue);
    }
  }

  render() {
    const {
      options,
      placeholder,
      errorMessage,
      label,
      name,
      required,
      readOnly
    } = this.props;

    return <div className="cf-form-dropdown">
      <label className="question-label" htmlFor={name}>
        {label || name} {required && <span className="cf-required">Required</span>}
      </label>
      {errorMessage && <span className="usa-input-error-message">{errorMessage}</span>}
      <Select
        id={name}
        value={this.state.value}
        options={options}
        onChange={this.onChange}
        placeholder={placeholder ? placeholder : "Select option"}
        clearable={false}
        noResultsText="Not an option"
        disabled={readOnly}
      />
    </div>;
  }
}

SearchableDropdown.propTypes = {
  errorMessage: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  options: PropTypes.array,
  readOnly: PropTypes.bool,
  required: PropTypes.bool,
  placeholder: PropTypes.string
};

export default SearchableDropdown;
