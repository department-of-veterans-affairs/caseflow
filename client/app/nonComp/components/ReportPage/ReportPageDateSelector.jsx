import React from 'react';
import { useController } from 'react-hook-form';
import PropTypes from 'prop-types';

import DateSelector from 'app/components/DateSelector';
import { marginTop } from 'app/hearings/components/details/style';

const ReportPageDateSelector = ({ name, label, control, errorMessage }) => {
  const { field } = useController({
    control,
    name,
    label
  });

  return (
    <fieldset style={{ marginTop: '35px' }}>
      <div key={`date-selector-container-${name}`}>
        <DateSelector
          name={`${name}_1`}
          key={`date-selector-${name}`}
          label={label}
          stronglabel
          onChange={(val) => {
            field.onChange(val);
          }}
          type="date"
          noFutureDates
          inputStyling={marginTop('0 !important')}
          errorMessage={errorMessage}
        />
      </div>
    </fieldset>
  );
};

ReportPageDateSelector.propTypes = {
  name: PropTypes.string,
  label: PropTypes.string,
  control: PropTypes.object,
  errorMessage: PropTypes.string
};

export default ReportPageDateSelector;
