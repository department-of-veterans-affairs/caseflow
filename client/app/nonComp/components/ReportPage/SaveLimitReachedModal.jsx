
import React from 'react';
import Modal from 'app/components/Modal';
import COPY from 'app/../COPY';
import PropTypes from 'prop-types';

export const SaveLimitReachedModal = (props) => {
  const { setShowLimitModal } = props;

  const handleCancel = () => {
    setShowLimitModal(false);
  };

  const submitForm = () => {
    // console.log('this will be updated in next ticket');
  };

  return (
    <Modal title={COPY.SAVE_LIMIT_REACH_TITLE}
      buttons={[
        { classNames: ['cf-modal-link', 'cf-btn-link'],
          name: 'Cancel',
          onClick: handleCancel
        },
        { classNames: ['usa-button'],
          name: 'Delete',
          onClick: submitForm
        },
        { classNames: ['usa-button', 'usa-button-secondary', 'cf_add_margin'],
          name: 'View saved searches',
          onClick: submitForm
        }
      ]}
      closeHandler={handleCancel}
    >
      <p> {COPY.SAVE_LIMIT_REACH_MESSAGE}</p>
    </Modal>);
};

SaveLimitReachedModal.propTypes = {
  setShowLimitModal: PropTypes.func.isRequired
};

export default SaveLimitReachedModal;
