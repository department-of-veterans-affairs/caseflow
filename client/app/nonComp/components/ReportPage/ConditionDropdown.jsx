import React, { useState } from 'react';
import SearchableDropdown from 'app/components/SearchableDropdown';
import { Controller } from 'react-hook-form';
import PropTypes from 'prop-types';

export const ConditionDropdown = ({ control, determineOptions, name }) => {
  let [disabled, setDisabled] = useState(false);

  const filteredOptions = determineOptions();

  const dropdownName = `${name}.condition`;

  return <Controller
    control={control}
    name={dropdownName}
    defaultValue={null}
    render={({ onChange, ...rest }) => (
      <SearchableDropdown
        {...rest}
        label="Variable"
        options={filteredOptions}
        readOnly={disabled}
        onChange={(valObj) => {
          setDisabled(true);
          onChange(valObj?.value);
        }}
        placeholder="Select a variable"
      />
    )}
  />;
};

ConditionDropdown.propTypes = {
  control: PropTypes.object,
  determineOptions: PropTypes.func,
  name: PropTypes.string
};
