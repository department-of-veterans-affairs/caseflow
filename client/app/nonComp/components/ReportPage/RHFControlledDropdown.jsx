import React from 'react';
import { Controller, useFormContext } from 'react-hook-form';
import PropTypes from 'prop-types';

import SearchableDropdown from 'app/components/SearchableDropdown';

const RHFControlledDropdown = ({ control, ...props }) => (
  <>
    <h2>{ props.header }</h2>
    <Controller
      control={control}
      name={props.name}
      render={({ onChange, ...rest }) => (
        <SearchableDropdown
          {...rest}
          name={props.name}
          label={props.label}
          options={props.options}
          searchable={false}
          onChange={(valObj) => {
            onChange(valObj?.value);
          }}
          required={props.required}
          optional={props.optional}
        />
      )}
    />
  </>
);

export const RHFControlledDropdownContainer = (props) => {
  const methods = useFormContext();

  return <RHFControlledDropdown {...methods} {...props} />;
};

RHFControlledDropdown.propTypes = {
  control: PropTypes.object,
  header: PropTypes.string,
  name: PropTypes.string,
  label: PropTypes.string,
  options: PropTypes.array,
  optional: PropTypes.bool,
  required: PropTypes.bool
};

export default RHFControlledDropdownContainer;
