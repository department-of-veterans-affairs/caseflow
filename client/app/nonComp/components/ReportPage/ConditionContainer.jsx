import React from 'react';
import { useWatch, useFormContext } from 'react-hook-form';
import { ConditionDropdown } from './ConditionDropdown';
import Link from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/Link';
import { DaysWaiting } from './Conditions/DaysWaiting';
import { DecisionReviewType } from './Conditions/DecisionReviewType';
import { IssueType } from './Conditions/IssueType';
import { IssueDisposition } from './Conditions/IssueDisposition';
import { Facility } from './Conditions/Facility';
import { Personnel } from './Conditions/Personnel';
import PropTypes from 'prop-types';

export const ConditionContainer = ({ control, index, remove }) => {

  const { watch, register } = useFormContext();

  const variableOptions = [
    { label: 'Days Waiting',
      value: 'daysWaiting',
      component: DaysWaiting },
    { label: 'Decision Review Type',
      value: 'decisionReviewType',
      component: DecisionReviewType },
    { label: 'Issue Type',
      value: 'issueType',
      component: IssueType },
    { label: 'Issue Disposition',
      value: 'issueDisposition',
      component: IssueDisposition },
    { label: 'Personnel',
      value: 'personnel',
      component: Personnel },
    { label: 'Facility',
      value: 'facility',
      component: Facility },
  ];

  const determineOptions = () => {
    let conds = watch('conditions');
    let selectedOptions = conds.map((cond) => cond.condition).filter((cond) => cond !== null);

    // personnel and facility are mutually exclusive
    if (selectedOptions.includes('facility')) {
      selectedOptions = selectedOptions.concat('personnel');
    } else if (selectedOptions.includes('personnel')) {
      selectedOptions = selectedOptions.concat('facility');
    }

    return variableOptions.filter((option) =>
      !selectedOptions.some((selectedOption) => option.value === selectedOption));
  };

  const name = `conditions.${index}`;

  const conditionsLength = useWatch({ name: 'conditions' }).length;
  const shouldShowAnd = (conditionsLength > 1) && (index !== (conditionsLength - 1));

  const selectedConditionValue = useWatch({ control, name: `${name}.condition` });

  const getConditionContent = () => {
    const selectedVariableOption = variableOptions.find((opt) => opt.value === selectedConditionValue);

    if (!selectedConditionValue || !selectedVariableOption) {
      return <div></div>;
    }

    if (selectedVariableOption.component) {
      const ConditionContent = selectedVariableOption.component;

      return <ConditionContent {...{ control, register, name }} />;
    }
  };

  return <div className="report-page-segment">
    <div className="cf-app-segment cf-app-segment--alt report-page-variable-condition" >
      <div className="report-page-variable-select">
        <ConditionDropdown {...{ control, determineOptions, name }} />
      </div>
      <div className="report-page-variable-content">{selectedConditionValue ? getConditionContent() : null} </div>
    </div>
    <Link onClick={() => remove(index)}>Remove condition</Link>
    {shouldShowAnd ? <div className="report-page-condition-and">AND</div> : null}
  </div>;
};

ConditionContainer.propTypes = {
  control: PropTypes.object,
  index: PropTypes.number,
  remove: PropTypes.func
};
