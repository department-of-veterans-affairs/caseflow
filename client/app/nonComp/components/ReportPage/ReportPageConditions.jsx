import React from 'react';
import { useFormContext, useFieldArray } from 'react-hook-form';

import { ConditionContainer } from './ConditionContainer';
import { personnelSchema } from './Conditions/Personnel';
import Button from 'app/components/Button';

import * as yup from 'yup';
import { daysWaitingSchema } from './Conditions/DaysWaiting';
import { decisionReviewTypeSchema } from './Conditions/DecisionReviewType';
import * as ERRORS from 'constants/REPORT_PAGE_VALIDATION_ERRORS';

const conditionOptionSchemas = {
  daysWaiting: daysWaitingSchema,
  decisionReviewType: decisionReviewTypeSchema,
  facility: yup.object(),
  issueDisposition: yup.object(),
  issueType: yup.object(),
  personnel: personnelSchema
};

export const conditionsSchema = yup.array().of(
  yup.lazy((value) => {
    return yup.object(
      { condition: yup.string().typeError(ERRORS.MISSING_CONDITION).
        oneOf(['daysWaiting', 'decisionReviewType', 'facility', 'issueDisposition', 'issueType', 'personnel']).
        required(),
      options: conditionOptionSchemas[value.condition]
      });
  })
);

export const ReportPageConditions = () => {
  const { control, watch } = useFormContext();
  const { fields, append, remove } = useFieldArray({
    control,
    name: 'conditions',
    defaultValues: [{ condition: '', options: {} }]
  });

  const watchFieldArray = watch('conditions');
  const controlledFields = fields.map((field, index) => {
    return {
      ...field,
      ...watchFieldArray[index]
    };
  });

  return (
    <div>
      <hr style={{ margin: '50px 0' }} />
      {/* Update margin depending on the presence of controlledField elements */}
      <h2 style={controlledFields.length ? { margin: '0' } : null}>Conditions</h2>
      {controlledFields.map((field, index) => {
        return <ConditionContainer key={field.id} {... { control, index, remove, field }} />;
      })}
      <Button
        disabled={watchFieldArray.length >= 5}
        onClick={() => append({ condition: '', options: {} })}>
      Add Condition</Button>
    </div>
  );
};
