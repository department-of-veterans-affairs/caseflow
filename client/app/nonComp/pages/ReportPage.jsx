import React, { useEffect } from 'react';
import { useController, useForm, FormProvider, useFormContext } from 'react-hook-form';
import { useDispatch, useSelector } from 'react-redux';
import { downloadReportCSV } from 'app/nonComp/actions/changeHistorySlice';
import { css } from 'glamor';
import PropTypes from 'prop-types';

import Button from 'app/components/Button';
import NonCompLayout from '../components/NonCompLayout';
import { conditionsSchema, ReportPageConditions } from '../components/ReportPage/ReportPageConditions';

import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { fetchUsers } from 'app/nonComp/actions/usersSlice';

import RHFControlledDropdownContainer from 'app/nonComp/components/ReportPage/RHFControlledDropdown';
import { timingSchema, TimingSpecification } from 'app/nonComp/components/ReportPage/TimingSpecification';

import Checkbox from 'app/components/Checkbox';
import RadioField from 'app/components/RadioField';

import { get } from 'lodash';

import {
  REPORT_TYPE_OPTIONS,
  RADIO_EVENT_TYPE_OPTIONS,
  RADIO_STATUS_OPTIONS,
  RADIO_STATUS_REPORT_TYPE_OPTIONS,
  SPECIFIC_STATUS_OPTIONS,
  SPECTIFIC_EVENT_OPTIONS
} from 'constants/REPORT_TYPE_CONSTANTS';
import * as ERRORS from 'constants/REPORT_PAGE_VALIDATION_ERRORS';

const buttonInnerContainerStyle = css({
  display: 'flex',
  gap: '32px',
});

const buttonOuterContainerStyling = css({
  display: 'flex',
  justifyContent: 'space-between',
  marginTop: '4rem',
});

const specificEventTypeSchema = yup.lazy((value) => {
  // eslint-disable-next-line no-undefined
  if (value === undefined) {
    return yup.mixed().notRequired();
  }

  return yup.object({
    added_decision_date: yup.boolean(),
    added_issue: yup.boolean(),
    added_issue_no_decision_date: yup.boolean(),
    claim_created: yup.boolean(),
    claim_closed: yup.boolean(),
    claim_status_incomplete: yup.boolean(),
    claim_status_inprogress: yup.boolean(),
    completed_disposition: yup.boolean(),
    removed_issue: yup.boolean(),
    withdrew_issue: yup.boolean(),
  }).test('at-least-one-true', ERRORS.AT_LEAST_ONE_OPTION, (obj) => {
    return Object.values(obj).some((val) => val === true);
  });
});

const specificStatusSchema = yup.lazy((value) => {
  // eslint-disable-next-line no-undefined
  if (value === undefined) {
    return yup.mixed().notRequired();
  }

  return yup.object({
    incomplete: yup.boolean(),
    in_progress: yup.boolean(),
    completed: yup.boolean(),
  }).test('at-least-one-true', ERRORS.AT_LEAST_ONE_OPTION, (obj) => {
    return Object.values(obj).some((val) => val === true);
  });
});

const schema = yup.object().shape({
  conditions: conditionsSchema,
  timing: timingSchema,
  specificEventType: specificEventTypeSchema,
  specificStatus: specificStatusSchema
});

const ReportPageButtons = ({
  history,
  isGenerateButtonDisabled,
  handleClearFilters,
  handleSubmit,
  loading,
  loadingText }) => {
  return (

    <div {...buttonOuterContainerStyling}>
      <Button
        classNames={['cf-modal-link', 'cf-btn-link']}
        label="cancel-report"
        name="cancel-report"
        onClick={() => history.push('/vha')}
      >
        Cancel
      </Button>
      <div {...buttonInnerContainerStyle}>
        <Button
          classNames={['usa-button-secondary']}
          label="clear-filters"
          name="clear-filters"
          onClick={handleClearFilters}
          disabled={isGenerateButtonDisabled}
        >
          Clear filters
        </Button>
        <Button
          classNames={['usa-button']}
          label="generate-report"
          name="generate-report"
          onClick={handleSubmit}
          disabled={isGenerateButtonDisabled}
          loading={loading}
          loadingText={loadingText}
        >
          Generate task report
        </Button>
      </div>
    </div>
  );
};

const RHFCheckboxGroup = ({ options, name, control }) => {
  const { field } = useController({
    control,
    name
  });

  const { errors } = useFormContext();

  const [value, setValue] = React.useState({});

  let fieldClasses = 'checkbox';

  const errorMessage = get(errors, name)?.message;

  if (errorMessage) {
    fieldClasses += ' usa-input-error';
    fieldClasses += ' less-error-padding';
  }

  return (
    <fieldset className={fieldClasses} style={{ paddingLeft: '30px' }}>
      {errorMessage ? <div className="usa-input-error-message">{ errorMessage }</div> : null}
      {options.map((option) => (
        <div key={option.id}>
          <Checkbox
            name={`${name}.${option.id}`}
            key={`${name}.${option.id}`}
            label={option.label}
            stronglabel
            onChange={(val) => {
              value[option.id] = val;
              field.onChange(value);
              setValue(value);
            }}
            unpadded
          />
        </div>
      ))}
    </fieldset>
  );
};

const RHFRadioButton = ({ options, name, control, label }) => {
  const { field } = useController({
    control,
    name,
  });

  return (
    <div style={{ marginTop: '20px' }}>
      <RadioField
        name=""
        label={label}
        vertical
        options={options}
        stronglabel
        value={field.value}
        onChange={(val) => {
          field.onChange(val);
        }}
      />
    </div>
  );
};

const ReportPage = ({ history }) => {
  const defaultFormValues = {
    reportType: '',
    conditions: [],
    timing: {
      range: null,
      startDate: '',
      endDate: '',
    },
    radioEventAction: 'all_events_action',
    radioStatus: 'all_statuses',
    radioStatusReportType: 'last_action_taken',
    specificStatus: {
      incomplete: '',
      in_progress: '',
      completed: '',
      cancelled: ''
    },
    specificEventType: {
      added_decision_date: '',
      added_issue: '',
      added_issue_no_decision_date: '',
      claim_created: '',
      claim_closed: '',
      claim_status_incomplete: '',
      claim_status_inprogress: '',
      completed_disposition: '',
      removed_issue: '',
      withdrew_issue: '',
    }
  };

  const methods = useForm({
    defaultValues: { ...defaultFormValues },
    resolver: yupResolver(schema),
    mode: 'onSubmit',
    reValidateMode: 'onSubmit'
  });

  const { reset, watch, formState, control, handleSubmit } = methods;
  const dispatch = useDispatch();
  const businessLineUrl = useSelector((state) => state.nonComp.businessLineUrl);
  const csvGeneration = useSelector((state) => state.changeHistory.status);
  const isCSVGenerating = csvGeneration === 'loading';
  const watchReportType = watch('reportType');
  const watchRadioEventAction = watch('radioEventAction');
  const watchRadioStatus = watch('radioStatus');

  const processConditionOptions = (condition, options) => {
    let formattedOptions;

    switch (condition) {
    case 'decisionReviewType':
      formattedOptions = Object.keys(options).filter((key) => options[key]);
      break;
    // Multi select conditions
    case 'personnel':
    case 'facility':
    case 'issueDisposition':
    case 'issueType':
      formattedOptions = Object.values(options)[0].map((item) => item.value);
      break;
    // Else it is probably already an object, so it just pass the existing options
    default:
      formattedOptions = options;
    }

    return formattedOptions;
  };

  const parseFilters = (data) => {
    const filters = {};

    // Add report type to the filter
    filters.reportType = data.reportType;

    // Event specific event types to the filter
    if (data.radioEventAction === 'specific_events_action') {
      filters.events = Object.keys(data.specificEventType).filter((key) => data.specificEventType[key] === true);
    }

    // Add specific status types to the filter
    if (data.radioStatus === 'specific_status') {
      filters.statuses = Object.keys(data.specificStatus).filter((key) => data.specificStatus[key] === true);
    }

    // Add timing to the filter
    filters.timing = data.timing;

    // Add Status report type to the filter
    filters.statusReportType = data.radioStatusReportType;

    // Parse all the Conditions and add them to the filter
    const transformedConditions = data?.conditions?.reduce((result, item) => {
      const { condition, options } = item;

      if (condition && options) {
        // Parse the individual conditions
        const newOptions = processConditionOptions(condition, options);

        result[condition] = newOptions;
      }

      return result;
    }, {});

    // Add the all the parsed conditions into the filters
    Object.assign(filters, transformedConditions);

    return filters;
  };

  const submitForm = (data) => {
    const filterData = parseFilters(data);

    // Generate and trigger the download of the CSV
    dispatch(downloadReportCSV({ organizationUrl: businessLineUrl, filterData: { filters: filterData } }));
  };

  useEffect(() => {
    dispatch(fetchUsers({ queryType: 'organization', queryParams: { query: 'vha' } }));
  }, []);

  return (
    <NonCompLayout
      buttons={
        <ReportPageButtons
          history={history}
          isGenerateButtonDisabled={!formState.isDirty}
          handleClearFilters={() => reset()}
          handleSubmit={handleSubmit(submitForm)}
          loading={isCSVGenerating}
          loadingText="Generating CSV"
        />
      }
    >
      <h1>Generate task report</h1>
      <FormProvider {...methods}>
        <form>
          <RHFControlledDropdownContainer
            header="Type of report"
            name="reportType"
            label="Report Type"
            options={REPORT_TYPE_OPTIONS}
          />
          {watchReportType === 'status' ? (<>
            <RHFRadioButton
              options={RADIO_STATUS_OPTIONS}
              methods={methods}
              name="radioStatus"
            />
            {watchRadioStatus === 'specific_status' ? (
              <RHFCheckboxGroup
                options={SPECIFIC_STATUS_OPTIONS}
                control={control}
                name="specificStatus"
              />) :
              null
            }
            <RHFRadioButton
              options={RADIO_STATUS_REPORT_TYPE_OPTIONS}
              methods={methods}
              label="Select type of status report"
              name="radioStatusReportType"
            />
          </>) :
            null
          }
          {watchReportType === 'event_type_action' ? (
            <RHFRadioButton
              options={RADIO_EVENT_TYPE_OPTIONS}
              methods={methods}
              name="radioEventAction"
            />
          ) : null
          }
          {watchReportType === 'event_type_action' &&
          watchRadioEventAction === 'specific_events_action' ? (
              <RHFCheckboxGroup
                options={SPECTIFIC_EVENT_OPTIONS}
                control={control}
                name="specificEventType"
              />
            ) : null
          }
          {watchReportType === 'event_type_action' ?
            <TimingSpecification /> :
            null
          }
          {formState.isDirty ? <ReportPageConditions /> : null}
        </form>
      </FormProvider>
    </NonCompLayout>
  );
};

ReportPageButtons.propTypes = {
  history: PropTypes.object,
  isGenerateButtonDisabled: PropTypes.bool,
  handleClearFilters: PropTypes.func,
  handleSubmit: PropTypes.func,
  loading: PropTypes.bool,
  loadingText: PropTypes.string,
};

ReportPage.propTypes = {
  history: PropTypes.object,
};

RHFCheckboxGroup.propTypes = {
  options: PropTypes.array,
  control: PropTypes.object,
  name: PropTypes.string,
  errorMessage: PropTypes.string
};

RHFRadioButton.propTypes = {
  options: PropTypes.array,
  control: PropTypes.object,
  label: PropTypes.string,
  name: PropTypes.string
};

export default ReportPage;
