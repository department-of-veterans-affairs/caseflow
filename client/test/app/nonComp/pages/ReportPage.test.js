import React from 'react';
import { axe } from 'jest-axe';

import userEvent from '@testing-library/user-event';
import { fireEvent, render, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import ReportPage from 'app/nonComp/pages/ReportPage';
import SearchableDropdown from 'app/components/SearchableDropdown';
import { shallow } from 'enzyme';

describe('ReportPage', () => {
  const setup = () => {
    return render(
      <ReportPage />
    );
  };

  it('passes a11y testing', async () => {
    const { container } = setup();

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('renders correctly', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('brings you to the decision review page when clicking the cancel button', async () => {
    const history = createMemoryHistory();

    render(
      <ReportPage history={history} />
    );

    const cancelButton = screen.getByText('Cancel');

    await userEvent.click(cancelButton);

    expect(history.location.pathname).toBe('/vha');
  });

  it('should have Generate task Report button and Clear Filter button disabled on initial load', () => {
    render(
      <ReportPage />
    );

    const generateTaskReport = screen.getByRole('button', { name: /Generate task Report/i });
    expect(generateTaskReport).toHaveClass('usa-button-disabled');

    const clearFilters = screen.getByText('Clear filters');
    expect(clearFilters).toHaveClass('usa-button-disabled');
  });
});
