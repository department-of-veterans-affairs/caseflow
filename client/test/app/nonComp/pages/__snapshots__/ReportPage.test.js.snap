// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportPage renders correctly 1`] = `
<div>
  <div>
    <section
      class="cf-app-segment cf-app-segment--alt"
    >
      <div>
        <h1>
          Generate task report
        </h1>
        <form>
          <h2>
            Type of Report
          </h2>
          <div
            class=""
          >
            <div
              class="cf-form-dropdown dropdown-reportType"
            >
              <label
                class="question-label"
                for="report-type"
                id="report-type-label"
              >
                <span>
                  Report Type
                </span>
              </label>
              <div
                class="cf-select"
              >
                <div
                  class=" css-2b097c-container"
                >
                  <div
                    class="cf-select__control css-yk16xz-control"
                  >
                    <div
                      class="cf-select__value-container css-g1d714-ValueContainer"
                    >
                      <div
                        class="cf-select__placeholder css-1wa3eu0-placeholder"
                      >
                        Select...
                      </div>
                      <div
                        class="css-81nyic-Input"
                      >
                        <div
                          class="cf-select__input"
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            aria-expanded="false"
                            aria-haspopup="true"
                            aria-labelledby="report-type-label"
                            aria-owns="report-type-listbox"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            id="report-type"
                            role="combobox"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="cf-select__indicators css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class="cf-select__indicator-separator css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class="cf-select__indicator cf-select__dropdown-indicator css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-6q0nyr-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <input
                    name="reportType"
                    type="hidden"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </section>
    <div
      data-css-1rfrm1q=""
    >
      <span>
        <button
          class="cf-modal-link cf-btn-link usa-button"
          id="button-cancel-report"
          type="button"
        >
          Cancel
        </button>
      </span>
      <div
        data-css-cfdlna=""
      >
        <span>
          <button
            class="usa-button-disabled usa-button"
            disabled=""
            id="button-clear-filters"
            type="button"
          >
            Clear filters
          </button>
        </span>
        <span>
          <button
            class="usa-button-disabled usa-button"
            disabled=""
            id="button-generate-report"
            type="button"
          >
            Generate task report
          </button>
        </span>
      </div>
    </div>
  </div>
</div>
`;
